resolwe/flow/migrations/0037_migrate_m2o.py

```diff
@@ -0,0 +1,227 @@
+
+    copy_permissions(original, duplicate, model='entity', apps=apps)
+
+    # Override fields that are automatically set on create.
+    duplicate.created = original.created
+    duplicate.save()
+
+    # Duplicate entity's data objects.
+    bundle = [{'original': obj, 'copy': duplicate_data(obj, apps)} for obj in original.data.all()]
+    bundle = rewire_inputs(bundle)
+    duplicated_data = [item['copy'] for item in bundle]
+
+    duplicate.data.add(*duplicated_data)
+
+    return duplicate
+
+
+def duplicate_data(original, apps):
+    """Duplicate (make a copy)."""
+    DataDependency = apps.get_model('flow', 'DataDependency')
+    Data = apps.get_model('flow', 'Data')
+
+    # TODO
+    # if original.status not in ['OK', 'ER']:
+    #     raise ValidationError('Data object must have done or error status to be duplicated')
+
+    duplicate = Data.objects.get(id=original.id)
+    duplicate.pk = None
+    duplicate.slug = None
+    duplicate.entity2 = None
+    duplicate.collection = None
+    duplicate.name = 'Copy of {}'.format(original.name)
+    duplicate.duplicated = now()
+
+    duplicate.save(force_insert=True)  # pylint: disable=protected-access
+
+    copy_permissions(original, duplicate, model='data', apps=apps)
+
+    # Override fields that are automatically set on create.
+    duplicate.created = original.created
+    duplicate.save()  # pylint: disable=protected-access
+
+    if original.location:
+        original.location.data.add(duplicate)  # pylint: disable=no-member
+
+    duplicate.storages.set(original.storages.all())  # pylint: disable=no-member
+
+    for migration in original.migration_history.order_by('created'):  # pylint: disable=no-member
+        migration.pk = None
+        migration.data = duplicate
+        migration.save(force_insert=True)
+
+    # Inherit existing child dependencies.
+    DataDependency.objects.bulk_create([
+        DataDependency(child=duplicate, parent=dependency.parent, kind=dependency.kind)
+        for dependency in DataDependency.objects.filter(child=original)
+    ])
+    # Inherit existing parent dependencies.
+    DataDependency.objects.bulk_create([
+        DataDependency(child=dependency.child, parent=duplicate, kind=dependency.kind)
+        for dependency in DataDependency.objects.filter(parent=original)
+    ])
+
+    return duplicate
+
+
+def duplicate_m2m(apps, schema_editor):
+    """Duplicate Data/Entity belonging to more than one Entity/Collection."""
+    Data = apps.get_model('flow', 'Data')
+    Entity = apps.get_model('flow', 'Entity')
+
+    # 1: Duplicate entities that are in multiple collections. This is
+    # done first as entities also contain data that is therefore
+    # handled "for free".
+    for entity in Entity.objects.annotate(collection_count=models.Count('collections')).filter(collection_count__gt=1):
+        collections = entity.collections.all()
+        for i, collection in enumerate(collections):
+            if i == 0:
```

Comment:
Why is this done here? It needs to be applied to all entities, not just those that are in multiple collections.

Reply:
Duplication has to be done only for the entities in multiple collections. Reassignment from m2m to m2o fields is done in `main`.